### ISSUE TITLE
color-rgb-only

### ISSUE THREAD

**Issue by merlinpatt** [stylelint/stylelint#2148]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A new rule (feature) that would only allow for rgb and rgba values for color

> What CSS is needed to reproduce this issue?

e.g.

```css
.foo {
  color: rgb(7, 7, 7);
}
```

> Which version of stylelint are you using?

`7.6.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

"CLI with `stylelint --config myconfig *.css`"

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No

**Comment by evilebottnawi** [comment]
@merlinpatt How about this http://stylelint.io/user-guide/rules/declaration-property-value-whitelist/ rule?

**Comment by jeddy3** [comment]
@merlinpatt Thanks for the request. What you're after can be achieved with:

- [`color-named`](http://stylelint.io/user-guide/rules/color-named/) and [`function-whitelist`](http://stylelint.io/user-guide/rules/function-whitelist/) (or `function-blacklist`]), if you want to do this for _all_ colours. 
- [`declaration-property-value-whitelist`](http://stylelint.io/user-guide/rules/declaration-property-value-whitelist/), as @evilebottnawi point out, if you want to do this for _just_ the `color` property.

The `*-blacklist` and `*-whitelist` rules are flexible enough, and work together, to enforce most desired conventions when it comes to allowing or disallowing language features.

**Comment by merlinpatt** [comment]
While the whitelist (or blacklist) will work, as wellas the regex options, this seems like something that should be easily declared within the configuration itself as its own rule. It also means that one needs an understanding of regex to configure that, which shouldn't be necessary to configure a linter.

Some of the configuration rules could in fact be done with regex (color-hex-case, color-named for example), yet these are still rules. What determines that these should be rules as opposed to other rules that can also be done via regex?

**Comment by davidtheclark** [comment]
> this seems like something that should be easily declared within the
configuration itself as its own rule

Can you please explain why you think rgb-only enforcement is important to
many users and not just a special case?

On Wed, Nov 30, 2016 at 10:42 AM, Merlin Patterson <notifications@github.com

**Comment by merlinpatt** [comment]
I don't have any research that would prove that this more than just a special case. But I do think that a not insignificant portion of developers use rgb, just as there are a non-insignificant number who use hex.

I'm unclear why there are multiple rules regarding hex values but none for rgb or hsl, which are also valid color values. From this, I infer that stylelint believes that developers should be using hex and not the other options, but linters should be neutral and provide multiple choices for the developers to make the decision on.

**Comment by davidtheclark** [comment]
> I infer that stylelint believes that developers should be using hex and not the other options

@merlinpatt: Honestly I think it's just because nobody has requested the others before.

@jeddy3: I think we need to address this request with something more tailored than a slightly imperfect combination of `color-named` and `function-whitelist` (which would still let hex values through, right?). How about a rule like `color-format` with options like `hex`, `rgb`, `rgba`, `hsl`, etc., for enforcing that only one format be used?

**Comment by merlinpatt** [comment]
I like the idea of a `color-format` choice. For future feature requests, I'll think about how to genericize the rule to be more of a choice.

**Comment by evilebottnawi** [comment]
@davidtheclark @merlinpatt 
`declaration-property-value-blacklist` with `"/.*/": ["/#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})/"]` - disable all properties use color hash

**Comment by davidtheclark** [comment]
@evilebottnawi Can't you see how that's not so user-friendly?

**Comment by evilebottnawi** [comment]
@davidtheclark but it is work :smile:

**Comment by evilebottnawi** [comment]
@davidtheclark 
> How about a rule like color-format with options like hex, rgb, rgba, hsl, etc., for enforcing that only one format be used?
Best solution is array -> ['hex', 'rgba', 'hls'], example i don't like `rgb` and `rgba` i use hex. But on the other hand, then it is necessary to mean all of the options which may contain colors and mean all possible functions, and it is not easy.

**Comment by evilebottnawi** [comment]
@merlinpatt you want use `rgb` only for `color` properties or other?

**Comment by merlinpatt** [comment]
I would like to use it for `color` related properties so `color`, `background-color`, etc

**Comment by evilebottnawi** [comment]
@merlinpatt why not use `"color-no-hex": true` and disable hex everywhere?

**Comment by davidtheclark** [comment]
So are we recommending:
- Disable hex values with `color-no-hex`
- Disable named colors with `color-no-named`
- Disable all color functions except `rgb` with a blacklist

Then maybe you can effectively enforce rgb-only.

That's kind of ridiculous for somebody with a fairly simple request, right?

**Comment by evilebottnawi** [comment]
@davidtheclark Perhaps, but we can create a document which will put such practices, thus save time when new request is comming.

**Comment by evilebottnawi** [comment]
@jeddy3 what do you think about notes? and we can close this issue?

**Comment by ntwb** [comment]
> How about a rule like `color-format` with options like `hex`, `rgb`, `rgba`, `hsl`, etc., for enforcing that only one format be used?

I'm +1 to the above...

**Comment by jeddy3** [comment]
We're touching on some original design decisions here. So, I figured this would be a good time to:

1. Document within the user guide how the (dis)allowing system works.
2. Recall within this issue some of those original design decisions. 

I've created a PR to [add a section to](https://github.com/stylelint/stylelint/pull/2189) the "About rules" doc.

I've tried to show that the design of our system provides an alternative to the need for, for example, specific rules to disallow `@debug` ([ref](https://github.com/brigade/scss-lint/tree/master/lib/scss_lint/linter#debugstatement)) or the `none` value for the `border` properties ([ref](https://github.com/brigade/scss-lint/tree/master/lib/scss_lint/linter#borderzero)).

I've then tried to show how that system scales up to use cases like wanting to enforce using a named color *if one exists for a color* and use `hwb` color if one does not. This hopefully illustrates our consistent use of explicit options and the advantages it brings.

And then show how the system can cope with language extensions (that use the two built-in extendable syntactic constructs of at-rules and functions).

> I'm unclear why there are multiple rules regarding hex values but none for rgb or hsl, which are also valid color values. From this, I infer that stylelint believes that developers should be using hex and not the other options, but linters should be neutral and provide multiple choices for the developers to make the decision on.

Simply because `rgb` and `hsl` extra are functions. There are 17 `function-*` rules, 11 of which are applicable to `rgb` and `hsl`. There are rules for allowing and disallowing which functions are allowed, specifying what case should be used, specifying spacing around punctuation etc. There are more rules applicable to color functions than there are for hex colours or named ones.

> I think we need to address this request with something more tailored than a slightly imperfect combination of color-named and function-whitelist (which would still let hex values through, right?)

Agreed. However, I hope to argue that our current system _is_ the best compromise between user experience, covering off the most number of use cases and keeping maintenance of the rules sustainable.

In the PR I think I cover most of how the (dis)allowing rules works from a user's perspective.  One thing I didn't touch on is avoiding conflicts. I think the only possible conflict, within the 170 rules, is if a `*-blacklist*` and `*-whitelist` for the _same construct_ are used e.g. `function-whitelist` and `function-blacklist`. (The conflicting `*-quotes` rules were resolved a few versions ago.) These conflicts are, hopefully, quite obvious to a user. However, adding `color-format` would introduce a less obvious conflict e.g.

```js
"function-whitelist": ["var", "scale"]
"color-format": ["rgb"]
```

Right, on with the recall... 

From the get-go we set out to build laser-focused rules because they are easier to maintain and easier to test. (We subsequently found out, when we came to build the whitespace rules, that these laser-focused rules could be [put together](http://stylelint.io/user-guide/about-rules/#-newlinespace-before-and--newlinespace-after) to enforce _a wide vareity_ of code styles).

Incidentally, it's also the reason we have individual `*-no-*` rules for each of the different types of selector, rather than something like a `allowed-selector-types: []" rule.

I think the `color-format` proposal will involve us munging together the code and tests of the `color-named`, `color-no-hex` `function-whitelist` rules into a single rule, which would go against our original design decisions. Decisions that have stood us very well.

Considering the breadth of the system (i.e. coverage of the language features and language extensions), how we've so far kept development on the rules sustainable, and how many use cases it covers, I think we've already found the best compromise.

**Comment by davidtheclark** [comment]
Sounds good to me. The documentation PR makes sense. I do think it's a somewhat painful UX, but you're right, @jeddy3, that we should make that sacrifice for the sake of controlling the core rules.

**Comment by what1s1ove** [comment]
It's sad that there's still nothing (at least no simple way) to restrict a project to use only a limited set of color formats.
The number of color formats keeps increasing, which means the need for such a rule is becoming greater.
